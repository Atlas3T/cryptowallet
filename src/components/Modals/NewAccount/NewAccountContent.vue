<template>
  <div>
    <div class="header-section">
      <div class="header-back-button-wrapper">
        <q-btn
          icon="arrow_back"
          size="lg"
          class="icon-btn back-arrow-btn"
          flat
          @click.prevent="closeModal"
        />
      </div>
      <h1 class="header-h1">
        {{ $t('newAccount') }}
      </h1>
    </div>

    <div class="modal-layout-wrapper centered">
      <div>
        <q-btn
          :label="$t('createAccount')"
          icon="add_box"
          color="primary"
          text-color="white"
          @click="createAccount"
        />
      </div>
      <div class="advanced-dropdown q-mt-sm">
        <q-expansion-item
          label="Advanced"
          color="info"
        >
          <q-btn
            :label="$t('seedPhraseCreate')"
            flat
            no-caps
            dense
            class="import-account-btn"
            text-color="blue"
            size="13px"
            @click="seedPhrase()"
          />
          <q-btn
            :label="$t('seedPhraseImport')"
            flat
            no-caps
            dense
            class="import-account-btn"
            text-color="blue"
            size="13px"
            @click="importAccount()"
          />
        </q-expansion-item>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'NewAccountContent',
  methods: {
    closeModal() {
      this.$store.dispatch('modals/setNewAccountModalOpened', false);
    },
    createAccount() {
      this.$store.dispatch('modals/setSelectAccountModalOpened', false);
      this.$store.dispatch('modals/setNewAccountModalOpened', false);
      this.$store.dispatch('setup/setAccountType', 'new');
      this.$router.push({ path: '/setup/8' });
    },
    importAccount() {
      this.$store.dispatch('modals/setSelectAccountModalOpened', false);
      this.$store.dispatch('modals/setNewAccountModalOpened', false);
      this.$store.dispatch('setup/setAccountType', 'restored');
      this.$router.push({ path: '/setup/1' });
    },

    seedPhrase() {
      this.$store.dispatch('setup/setAccountLocale', this.selectedLang.value);
      this.$store.dispatch('setup/setAccountCurrency', this.$t('supportedCurrency'));
      this.$store.dispatch('setup/setAccountType', 'new');
      this.$router.push({ path: '/setup/2' });
    },
  },
};
</script>

<style>
.import-account-btn {
  width: 15rem;
}
.advanced-dropdown .q-expansion-item__container {
  position: absolute;
  left: 50%;
  transform: translate(-50%);
}

.advanced-dropdown .q-item {
  justify-content: center;
  min-height: 0px;
}

.advanced-dropdown .q-item__section--main {
  flex: none;
}

.advanced-dropdown .q-item__section--main ~ .q-item__section--side {
  padding-left: 2px;
}
</style>
