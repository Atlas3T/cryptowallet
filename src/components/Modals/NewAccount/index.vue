<template>
  <div>
    <q-dialog
      v-model="newAccountModalOpened"
      persistent
      :maximized="true"
      transition-show="slide-up"
      transition-hide="slide-down"
      content-class="dark-modal"
    >
      <div class="header-section">
        <div class="header-back-button-wrapper">
          <q-btn
            icon="arrow_back"
            size="lg"
            class="icon-btn back-arrow-btn"
            flat
            @click.prevent="closeModal"
          />
        </div>
        <h1 class="header-h1">
          {{ $t('newAccount') }}
        </h1>
      </div>

      <div class="modal-layout-wrapper centered">
        <div>
          <q-btn
            :label="$t('createAccount')"
            icon="add_box"
            color="primary"
            text-color="blueish"
            @click="createAccount"
          />
        </div>
        <div class="btns-wrapper">
          <q-btn
            :label="$t('importAccount')"
            icon="get_app"
            color="primary"
            text-color="blueish"
            @click="importAccount"
          />
        </div>
      </div>
    </q-dialog>
  </div>
</template>

<script>
export default {
  name: 'NewAccount',
  data() {
    return {
      newAccountModalOpened: false,
    };
  },
  mounted() {
    this.$root.$on('newAccountModalOpened', (value) => {
      this.newAccountModalOpened = value;
    });
  },
  methods: {
    closeModal() {
      this.$root.$emit('newAccountModalOpened', false);
    },
    createAccount() {
      // @todo, don't use app global, should work
      /* eslint-disable-next-line */
      app.$root.$emit('selectAccountModalOpened', false);
      this.$root.$emit('newAccountModalOpened', false);
      this.$store.dispatch('setup/setAccountType', 'new');
      this.$router.push({ path: '/setup/2' });
    },
    importAccount() {
      // @todo, don't use app global, should work
      /* eslint-disable-next-line */
      app.$root.$emit('selectAccountModalOpened', false);
      this.$root.$emit('newAccountModalOpened', false);
      this.$store.dispatch('setup/setAccountType', 'restored');
      this.$router.push({ path: '/setup/1' });
    },
  },
};
</script>

<style>

</style>
