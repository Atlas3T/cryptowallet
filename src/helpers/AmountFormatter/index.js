/**
 * @TODO James Konrad
 * I've mocked some exchange rates below for the amount() function to use
 * Would make sense to put them somewhere in the Vuex store
 */

import numeral from 'numeral';

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GBP',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GBP',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GBP',
      rate: 9999,
    },
  ],
};

export default class AmountFormatter {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  getFormatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${this.currency.symbol}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `- ${formatted}`;
      if (amount > 0) return `+ ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency.code);
    if (rate) return this.amount * rate;
    return this.amount;
  }
}
