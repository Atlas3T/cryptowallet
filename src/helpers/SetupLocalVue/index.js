import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import VueI18n from 'vue-i18n';
import messages from '@/i18n/';
import axios from 'axios';
import { sync } from 'vuex-router-sync';
import errorHandlerPlugin from '@/boot/ErrorHandler';
import toasterPlugin from '@/boot/Toaster';
import permissionsPlugin from '@/boot/Permissions';
import appInvitationPlugin from '@/boot/AppInvitation';
import smsPlugin from '@/boot/Sms';
import emailPlugin from '@/boot/Email';
import contactsImportPlugin from '@/boot/ContactsImport';
import * as All from 'quasar'

const { Quasar, date } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})


const localVue = createLocalVue();

localVue.use(Quasar, { components });
localVue.use(VueI18n);
localVue.use(Vuex);
// localVue.use(VueRouter);
localVue.prototype.$axios = axios;
errorHandlerPlugin({ Vue: localVue });
toasterPlugin({ Vue: localVue });
permissionsPlugin({ Vue: localVue });
appInvitationPlugin({ Vue: localVue });
smsPlugin({ Vue: localVue });
emailPlugin({ Vue: localVue });
contactsImportPlugin({ Vue: localVue });

/**
 * creates a new router and syncs it with the store using vuex-router-sync
 * @param store
 * @returns router
 */
const createRouter = (store = null) => {
  const router = new VueRouter({
    routes: [
      {
        path: '/contact/:id',
      },
    ],
  });

  if (store) { sync(store, router); }
  return router;
};

const i18n = new VueI18n({ messages, locale: 'en-gb' });

export {
  localVue,
  createRouter,
  i18n,
};
