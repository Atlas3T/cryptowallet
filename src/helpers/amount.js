/**
 * @TODO James Konrad
 * mocked data below
 */

import numeral from 'numeral';

const currencySymbols = {
  USD: '$',
  GPB: 'Â£',
};

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GPB',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GPB',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GPB',
      rate: 9999,
    },
  ],
};

export default class Amount {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  get formatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${currencySymbols[this.currency]}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `&#45; ${formatted}`;
      if (amount > 0) return `&#43; ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency);
    return this.amount * rate;
  }
}
