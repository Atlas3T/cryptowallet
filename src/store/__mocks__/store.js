import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
