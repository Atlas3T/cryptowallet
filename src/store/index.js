/*eslint-disable*/

import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from 'vuex-orm-lokijs';

// import entities.
import Account from './wallet/entities/account';
import Address from './wallet/entities/address';
import Tx from './wallet/entities/tx';
import Utxo from './wallet/entities/utxo';
import Wallet from './wallet/entities/wallet';
import Coin from './wallet/entities/coin';
import KeyPair from './wallet/entities/keyPair';
import Prices from './prices';
import LatestPrice from './latestPrice';

// import modules.
import payments from './payments';
import search from './search';
import settings from './settings';
import setup from './setup';
import contacts from './contacts';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
];

window.Tx = Tx;
window.Utxo = Utxo;
window.Address = Address;
window.Wallet = Wallet;
window.LatestPrice = LatestPrice;
window.Prices = Prices;
window.Account = Account;
window.Coin = Coin;

// Setup persistant storage.
const vuexLocal = new VuexPersistence({
  supportCircular: true,
  storage: window.localStorage,
  modules: [settings],
});


const options = {
  env: 'browser',
};

function hydrationCompletedCallback() {
  setTimeout(() => {
    store.dispatch('settings/setLoading', false);
  }, 1000);
}

VuexORM.use(VuexORMLoki, { database, options, hydrationCompletedCallback });


/**
 * Create CryptoWallet Vuex store obj.
 * @type {Store<any>}
 */
const store = new Vuex.Store({
  modules: {
    payments,
    search,
    settings,
    setup,
    contacts,
  },
  plugins: [VuexORM.install(database)],
});

Vue.prototype.activeWallets = {};

// @todo remove line below (used for development purposes only)
window.store = store;

export default store;
