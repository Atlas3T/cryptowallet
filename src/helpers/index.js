import AmountFormatter from './AmountFormatter';
import dateTranslater from './DateTranslater';
import Contacts from './Contacts';

export {
  AmountFormatter,
  dateTranslater,
  Contacts,
};
