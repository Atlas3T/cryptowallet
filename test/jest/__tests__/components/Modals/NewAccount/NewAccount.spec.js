/* eslint-disable no-magic-numbers */
import { mount } from '@vue/test-utils';
import NewAccount from '@/components/Modals/NewAccount/NewAccountContent';
import NewAccountModal from '@/components/Modals/NewAccount';

import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';

describe('NewModalContent component', () => {
  let wrapper;
  let router;
  let storeMocks;

  const defaultProps = {};

  function wrapperInit(options) {
    return mount(NewAccount, options);
  }

  function storeInit(custom, propsData) {
    storeMocks = createStoreMocks(custom);
    router = createRouter(storeMocks.store);
    router.push({ path: '/' });
    wrapper = wrapperInit({
      i18n,
      router,
      localVue,
      propsData,
      parentComponent: NewAccountModal,
      store: storeMocks.store,
    });
  }

  beforeEach(() => { storeInit({}, defaultProps); });

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders create account and import account buttons', () => {
    expect(wrapper.findAll('button').at(1).text()).toMatch('Create Account');
    expect(wrapper.findAll('button').at(2).text()).toMatch('Import Account');
  });

  it('back button closes the modal', () => {
    wrapper.findAll('button').at(0).trigger('click');
    expect(storeMocks.actions.setNewAccountModalOpened).toHaveBeenCalled();
  });

  it('create account button closes modals and routes /setup/2', () => {
    wrapper.findAll('button').at(1).trigger('click');
    expect(storeMocks.actions.setNewAccountModalOpened).toHaveBeenCalled();
    expect(storeMocks.actions.setSelectAccountModalOpened).toHaveBeenCalled();
    expect(storeMocks.actions.setAccountType.mock.calls[0][1]).toBe('new');
    expect(router.history.current.path).toBe('/setup/2');
  });

  it('import account button closes modals and routes /setup/1', () => {
    wrapper.findAll('button').at(2).trigger('click');
    expect(storeMocks.actions.setNewAccountModalOpened).toHaveBeenCalled();
    expect(storeMocks.actions.setSelectAccountModalOpened).toHaveBeenCalled();
    expect(storeMocks.actions.setAccountType.mock.calls[1][1]).toBe('restored');
    expect(router.history.current.path).toBe('/setup/1');
  });

  describe('parent modal', () => {
    it('opens and closes the modal', () => {
      wrapper.vm.$parent.$store = wrapper.vm.$store;
      wrapper.vm.$parent.newAccountModalOpened = false;
      expect(wrapper.vm.$parent.newAccountModalOpened).toBe(false);
      expect(storeMocks.actions.setNewAccountModalOpened.mock.calls[3][1]).toBe(false);
      wrapper.vm.$parent.newAccountModalOpened = true;
      expect(storeMocks.actions.setNewAccountModalOpened.mock.calls[4][1]).toBe(true);
    });
  });
});
