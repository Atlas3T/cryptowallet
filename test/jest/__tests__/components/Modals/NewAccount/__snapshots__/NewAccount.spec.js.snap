// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewModalContent component renders and matches snapshot 1`] = `
<div>
  <div
    class="header-section"
  >
    <div
      class="header-back-button-wrapper"
    >
      <button
        class="q-btn inline q-btn-item non-selectable icon-btn back-arrow-btn q-btn--flat q-btn--rectangle q-focusable q-hoverable"
        style="font-size: 20px;"
        tabindex="0"
        type="button"
      >
        <div
          class="q-focus-helper"
          tabindex="-1"
        />
        <div
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="material-icons q-icon"
          >
            arrow_back
          </i>
        </div>
      </button>
    </div>
     
    <h1
      class="header-h1"
    >
      
      New Account
    
    </h1>
  </div>
   
  <div
    class="modal-layout-wrapper centered"
  >
    <div>
      <button
        class="q-btn inline q-btn-item non-selectable q-btn--standard q-btn--rectangle bg-primary text-blueish q-focusable q-hoverable"
        tabindex="0"
        type="button"
      >
        <div
          class="q-focus-helper"
          tabindex="-1"
        />
        <div
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="material-icons q-icon on-left"
          >
            add_box
          </i>
          <div>
            Create Account
          </div>
        </div>
      </button>
    </div>
     
    <div
      class="btns-wrapper"
    >
      <button
        class="q-btn inline q-btn-item non-selectable q-btn--standard q-btn--rectangle bg-primary text-blueish q-focusable q-hoverable"
        tabindex="0"
        type="button"
      >
        <div
          class="q-focus-helper"
          tabindex="-1"
        />
        <div
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="material-icons q-icon on-left"
          >
            get_app
          </i>
          <div>
            Import Account
          </div>
        </div>
      </button>
    </div>
  </div>
</div>
`;
